pkg/ccl/importccl/import_pgdump.go

```diff
@@ -0,0 +1,503 @@
+)
+
+// createTempImportDatabase creates a temporary database where we will create
+// all the PGDUMP objects during the import.
+func createTempImportDatabase(ctx context.Context, p sql.JobExecContext) (descpb.ID, error) {
+	id, err := catalogkv.GenerateUniqueDescID(ctx, p.ExecCfg().DB, p.ExecCfg().Codec)
+	if err != nil {
+		return 0, err
+	}
+	// TODO(adityamaru): Figure out how to create the database descriptor with
+	// privileges for only the node user. Currently, root and admin have ALL
+	// privileges on the database descriptor. This is enforced by
+	// `ValidateSuperuserPrivileges` when writing the descriptor to store.
+	// I tried moving the database descriptor to OFFLINE instead of mucking with
+	// privileges, but this prevents us from running any DDL statements on this
+	// database.
+	tempDBDesc := dbdesc.NewInitial(id, importTempPgdumpDB, security.NodeUserName())
+	return tempDBDesc.GetID(), sql.DescsTxn(ctx, p.ExecCfg(), func(ctx context.Context, txn *kv.Txn,
+		col *descs.Collection) error {
+		b := txn.NewBatch()
+		if err := col.WriteDescToBatch(
+			ctx, false /* kvTrace */, tempDBDesc, b,
+		); err != nil {
+			return err
+		}
+		b.CPut(catalogkeys.MakeDatabaseNameKey(p.ExecCfg().Codec, importTempPgdumpDB), tempDBDesc.GetID(), nil)
+		return txn.Run(ctx, b)
+	})
+}
+
+type postgresDDLHandler struct {
+	dumpDatabaseName string
+	// TODO(adityamaru): Maybe memory monitor?
+	bufferedDDLStmts []string
+}
+
+func formatPostgresStatement(n tree.NodeFormatter) string {
+	f := tree.NewFmtCtx(
+		// TODO(adityamaru): should this be serializable?
+		tree.FmtParsable,
+	)
+	f.FormatNode(n)
+	return f.CloseAndGetString()
+}
+
+func rewritePostgresStatementTableName(
+	n tree.NodeFormatter, dumpDatabaseName string,
+) (string, error) {
+	var err error
+	f := tree.NewFmtCtx(
+		// TODO(adityamaru): should this be serializable?
+		tree.FmtParsable,
+		tree.FmtReformatTableNames(func(ctx *tree.FmtCtx, tn *tree.TableName) {
+			// If the node has an explicit catalog name then we must replace it with
+			// the temporary database we are importing into.
+			if tn.CatalogName != "" {
+				if tn.CatalogName != tree.Name(dumpDatabaseName) {
+					err = errors.AssertionFailedf("catalog name %s does not match dump target database name %s",
+						tn.CatalogName, dumpDatabaseName)
+					return
+				}
+				tn.CatalogName = importTempPgdumpDB
+			}
+			// TODO (adityamaru): Is it possible for dump files to have db.object names?
+			// In that case, if the node has an explicit schema name, then this could
+			// be a schema name or catalog name. If the SchemaName matches the target
+			// database the dump file specified via a CREATE DATABASE statement, then
+			// we replace it with the temporary database we are importing into.
+			// Otherwise we leave it as is since it is a schema name. What if we have a
+			// schema with the same name as the target database?
+			// I'm not sure this is an issue so leaving it as a TODO for now.
```

Comment:
Note for our review meeting:

It is perhaps worth discussing how accommodating we need or want to be for input that postgresql ought not generate but which would still be parsed by us if we handled it.  I'm specifically thinking about the kind of cases you mention here where on the CRDB side we support referencing another database (in `CREATE TABLE` and `CREATE SCHEMA`) whereas on the postgresql side you can't reference another database as far as I know.

It _seems_ to me like in a number of cases below in bufferDDLPostgresStatement are also handling rename cases where we must be dealing with input not actually generated by pg_dump (although there might be something I've overlooked with respect to how some of these tree nodes are populated).  

I'm +1 on handling some of these cases since hand-editing pgdump files isn't uncommon, just might be good to spell out the bounds of what we want to support.